[STEP 3] 로또 자동 리뷰 요청 드립니다.

badaelephant opened the pull request:
안녕하세요 근환님 이번에도 미션이 진짜 재밌었습니다. 

이번에는 그래도 테스트 코드를 좀 먼저 작성하고 기능을 구현하려고 시도했습니다. 

최대한 일급컬렉션이나, 원시객체 포장, 그리고 기능을 구분하려고 노력했어요. 

그리고, 전에 재성님이 말씀해주신 getter를 사용하지 않으려고 시도를 했는데, 이러한 측면에서 부족한 부분 있으면 

가감없이 말씀해주세요 ㅎㅎ 감사합니다. 

좋은 주말 되세요~

Commit:
docs: 기능명세서 작성, 포맷 변경 (checkbox 가능하도록)

Commit:
test: 로또 번호 테스트코드 작성

Commit:
feat: 로또 번호 기능 구현, 입력 검증 유틸 클래스 생성

Commit:
test: 당첨 로또 기능 테스트 구현
사용자 입력 로또 승리 케이스

Commit:
feat: 당첨 로또 기능 구현
로또 안에 해당 로또번호가 있는지 확인하는 isMatched 기능

Commit:
fix: 로또 테스트 로또 번호포함 테스트 추가

Commit:
feat: 메시지 및 constant 클래스 생성

Commit:
test: 티켓 테스트 및 티켓리스트 테스트 생성 (자동생성된 로또번호)

Commit:
test: 결과 및 결과맵 테스트 생성

Commit:
feat: 티켓, 결과 기능 생성

Commit:
feat: 셔플 로또티켓 생성기 테스트 생성

Commit:
feat: 셔플 로또티켓 생성기 인터페이스 및 기능 구현

Commit:
feat: 돈을 입력하여 티켓 생성기를 통해 티켓을 생성하는 티켓머신 기능 생성

Commit:
feat: input과 output 기능 구현

Commit:
feat: 로또 어플리케이션 및 컨트롤러 구현

Commit:
refactor: 불필요한 임포트 삭제, 결과평가 테스트코드 및 기능 일부 수정

Commit:
docs: 리드미 기능명세서 업데이트

geunhwanlee commented on README.md:
```diff
@@ -6,4 +6,26 @@
\ No newline at end of file
+* [텍스트와 이미지로 살펴보는 온라인 코드 리뷰 과정](https://github.com/next-step/nextstep-docs/tree/master/codereview)
+
+## Step 3
```
기능 목록 정리 👍🏻

geunhwanlee commented on src/main/java/step3/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    public int hashCode() {
+        return Objects.hash(lotto);
+    }
```
1. equals와 hashCode를 오버라이딩한다는 것은 무엇을 의미할까요?
2. 또한 Lotto에서 equals와 hashCode를 오버라이딩한 것은 무엇을 의미할까요?

badaelephant replied:
Lotto의 경우 순서와 상관 없이 같음 으로 판단하는 것으로 로직을 추가하면 좋을 것 같습니다. 이를 통해서 빠르게 1등 티켓을 파악할 수도 있을것 같습니다.

geunhwanlee commented on src/main/java/step3/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,55 @@
+    public Lotto(String input) {
+        InputValidation.validateEmpty(input);
+        List<String> lottoNumbers = splitLottoNumbers(input);
+        this.lotto = createLotto(lottoNumbers);
```
final 선언과
```suggestion
        this.lotto = Collections.unmodifiableList(createLotto(lottoNumbers));
```
를 통해 불변 List를 만들 수 있어요.

참고하실만한 링크 보내드릴게요
https://ecsimsw.tistory.com/795

badaelephant replied:
불변리스트라는 것은 처음 들어봤는데, 알려주셔서 감사합니다. !!
확실히 기억하고, 다음부턴 변하면 안되는 컬렉션의 경우 꼭 적용하도록 하겠습니다.

geunhwanlee commented on src/main/java/step3/model/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,44 @@
+        InputValidation.validateEmpty(input);
+        InputValidation.validateNumber(input);
+        int number = Integer.parseInt(input);
+        InputValidation.validateIntRange(number);
```
유효성 검사 로직들을 유틸리티 클래스로 모아둔 이유가 있을까요?🤔
만약 유효성 검사 로직이 변경된다면 그 영향 범위는 얼마나 될까요?

badaelephant replied:
이부분도 알려주셔서 다시한번 깊게 생각해볼 수 있었는데요! 
예전에는 여러곳에서 사용하는 것은 공통 모듈로 빼는 습관이 있었는데요, 생각해보니 비슷한 로직이 여러군데 쓰이고 있다면, 그것은 객체 자체를 다시 그루핑해야한다는 것을 생각하게 되었습니다. 그리고 그 안으로 특화된 유효성 검사로직을 넣는게 좋겠습니다. 
정말 전역적으로 공통적으로 쓸, 변하지 않을 내용만 util로 빼는게 좋을것 같아요.

geunhwanlee commented on src/main/java/step3/model/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,44 @@
+    }
+    public LottoNumber(final int input) {
+        this.number = input;
+    }
```
int 타입을 받는 생성자는 유효성검사를 하지 않고있네요!

badaelephant replied:
수정했습니다. 감사합니다.

geunhwanlee commented on src/main/java/step3/model/lotto/Ticket.java:
```diff
@@ -0,0 +1,24 @@
+        this.ticket = ticket.stream().map(LottoNumber::new).collect(Collectors.toList());
+    }
+    public Result getResult(Lotto lotto){
+        int count = (int) ticket.stream().filter(number -> lotto.isMatched(number)).count();
```
```suggestion
        int count = (int) ticket.stream().filter(lotto::isMatched).count();
```
이렇게 쓸 수도 있겠네요.😊

badaelephant replied:
람다 사용이 좀 익숙하지 않았는데 알려주신 덕분에 효율적으로 적용할 수 있게 되었어요

geunhwanlee commented on src/main/java/step3/model/lotto/Ticket.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public String toString() {
+        return ticket.stream().map(LottoNumber::toString).collect(Collectors.toList()).toString();
```
```suggestion
        return ticket.toString();
```
이렇게 써도 되지 않을까요?

geunhwanlee commented on src/main/java/step3/model/lotto/Tickets.java:
```diff
@@ -0,0 +1,43 @@
+import step3.model.machine.Results;
+import step3.view.OutputView;
+
+public class Tickets {
```
일급컬렉션 👍🏻

geunhwanlee commented on src/main/java/step3/model/lotto/Tickets.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    private void printTickets() {
+        tickets.stream().forEach(ticket -> OutputView.printTicket(ticket));
```
도메인 계층인 Tickets에서 뷰 계층인 OutputView를 참조하는 것으로 보이네요.

Tickets를 View로 출력하는 역할은 어디에 있어야 할까요?🤔

geunhwanlee commented on src/main/java/step3/model/lotto/Tickets.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public int hashCode() {
+        return Objects.hash(tickets);
+    }
```
Tickets에서 equals와 hashCode는 어떤 목적으로 Override 하신걸까요?🤔

geunhwanlee commented on src/main/java/step3/model/lotto/Tickets.java:
```diff
@@ -0,0 +1,43 @@
+
+    public Results getResults(Lotto lotto) {
+        Results results = new Results();
+        tickets.stream().forEach(ticket -> results.recordResult(ticket.getResult(lotto)));
```
```suggestion
        tickets.forEach(ticket -> results.recordResult(ticket.getResult(lotto)));
```

geunhwanlee commented on src/main/java/step3/model/machine/Money.java:
```diff
@@ -0,0 +1,16 @@
+
+import step3.model.util.InputValidation;
+
+public class Money {
```
원시값 포장 👍🏻

geunhwanlee commented on src/main/java/step3/model/lotto/Ticket.java:
```diff
@@ -0,0 +1,24 @@
+import step3.model.machine.Result;
+import step3.model.util.InputValidation;
+
+public class Ticket {
```
Lotto와 Ticket은 서로 다른 여섯개의 숫자로 이루어진 하나의 로또라는 역할을 가지는 점에서 유사한 점이 많은 것 같은데요
Lotto와 Ticket으로 분리하여 작성한 이유가 있을까요?🤔

정답이 있는건 아니니 용석님의 생각을 말씀해 주셔도 됩니다!

geunhwanlee commented on src/main/java/step3/model/machine/Results.java:
```diff
@@ -0,0 +1,62 @@
+        lottoResultMap.put(result, lottoResultMap.get(result) + 1);
+    }
+
+    public Double evaluateResult(int ticketCount, long winningPrize) {
```
winningPrize는 Results가 알고있는 값인걸로 보이는데
파라미터로 받지 않아도 될것 같아요!

geunhwanlee commented on src/main/java/step3/model/machine/Results.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    public long getWinningPrize() {
+        return lottoResultMap.keySet().stream().mapToLong(result -> getTotalPrize(result)).sum();
```
```suggestion
        return lottoResultMap.keySet().stream().mapToLong(this::getTotalPrize).sum();
```
이렇게도 써볼 수 있겠네요

geunhwanlee commented on src/main/java/step3/model/machine/Results.java:
```diff
@@ -0,0 +1,62 @@
+                        result.toString(), Integer.toString(lottoResultMap.get(result))));
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof Results)) {
+            return false;
+        }
+        Results results = (Results) o;
+        return Objects.equals(lottoResultMap, results.lottoResultMap);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(lottoResultMap);
+    }
```
Results에서 equals와 hashCode를 Override한 이유는 무엇일까요?

geunhwanlee commented on src/main/java/step3/model/machine/Results.java:
```diff
@@ -0,0 +1,62 @@
+        return lottoResultMap.keySet().stream().mapToLong(result -> getTotalPrize(result)).sum();
+    }
+
+    public void printResults() {
```
printRresults가 의도와 다르게 동작하는 것 같습니다.
확인 한번 부탁드려요!😊

geunhwanlee commented on src/main/java/step3/model/machine/Results.java:
```diff
@@ -0,0 +1,62 @@
+    public void printResults() {
+        lottoResultMap.keySet().stream()
+                .filter(result -> result.isRewarded())
+                .forEach(result -> OutputView.printResult(
```
도메인 계층인 Results에서 뷰 계층인 OutputView를 참조하고 있네요.

Result를 뷰로 출력하는 역할은 어디에 있어야 할까요?🤔

badaelephant replied:
이부분도 사실 더 질문 드리고 싶은 것이 많았는데요, 
getter를 최대한 안쓰다보니, output에서 객체의 값을 효율적으로 가져오는 방법을 잘 모르겠더라구요. 
우선 변경방법은, return 을 특정 구분자로 구분되는 값을 ex) "4,5000,3" 전달하여 이를 split해서 OutputView에서 형태에 맞게 출력하도록 햇는데요. 혹시 더 좋은 방법이 있으면 알려주시면 감사하겠습니다.

geunhwanlee commented on src/main/java/step3/model/machine/ShuffleTicketGenerator.java:
```diff
@@ -0,0 +1,22 @@
+                .boxed()
+                .collect(toList());
+        Collections.shuffle(numbers);
+        List<Integer> ticket = numbers.subList(0, count);
+        return ticket;
```
```suggestion
        return numbers.subList(0, count);
```

geunhwanlee commented on src/main/java/step3/model/machine/ShuffleTicketGenerator.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.stream.IntStream;
+import step3.model.value.Rule;
+
+public class ShuffleTicketGenerator implements TicketGenerator {
```
랜덤성이 있는 로직을 ShuffleTicketGenerator로 분리한 것 좋습니다!👍🏻

geunhwanlee commented on src/main/java/step3/model/machine/TicketGenerator.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.List;
+import step3.model.lotto.Ticket;
+
+public interface TicketGenerator {
```
인터페이스 👍🏻

geunhwanlee commented on src/main/java/step3/model/machine/TicketMachine.java:
```diff
@@ -0,0 +1,24 @@
+import step3.model.value.Rule;
+import step3.view.OutputView;
+
+public class TicketMachine {
```
TicketMachine에 대한 테스트도 작성해 보는건 어떨까요?

geunhwanlee commented on src/main/java/step3/model/machine/TicketMachine.java:
```diff
@@ -0,0 +1,24 @@
+
+public class TicketMachine {
+    private final TicketGenerator ticketGenerator;
+    public TicketMachine(TicketGenerator ticketGenerator) {
```
의존성 주입 👍🏻

geunhwanlee commented on src/main/java/step3/model/machine/TicketMachine.java:
```diff
@@ -0,0 +1,24 @@
+        for (int i = 0; i < ticketCount; i++) {
+            tickets.add(new Ticket(ticketGenerator.generateTicket(Rule.LOTTO_NUMBER_LENGTH)));
+        }
+        OutputView.printTicketCount(ticketCount);
```
도메인 계층에서 뷰 계층을 의존하고 있네요!

geunhwanlee commented on src/main/java/step3/model/util/InputValidation.java:
```diff
@@ -0,0 +1,47 @@
+import step3.model.value.ErrMsg;
+import step3.model.value.Rule;
+
+public class InputValidation {
```
인스턴스 생성을 막기위해 private 생성자를 만들어 두는건 어떨까요?

geunhwanlee replied:
참고하실만한 링크 남겨드릴게요!
[인스턴스화를 막으려거든 private 생성자를 사용하라](https://velog.io/@lychee/이펙티브-자바-아이템-4.-인스턴스화를-막으려거든-private-생성자를-사용하라)

geunhwanlee commented on src/main/java/step3/model/util/InputValidation.java:
```diff
@@ -0,0 +1,47 @@
+public class InputValidation {
+    private static final String NUMBER_REGEX = "^[0-9]+$";
+
+    public static void validateNumber(String input){
+        if(!input.matches(NUMBER_REGEX)){
+            throw new IllegalArgumentException(ErrMsg.NOT_NUMBER);
+        }
+    }
+    public static void validateEmpty(String input){
+        if(input == null || input.isEmpty()){
+            throw new IllegalArgumentException(ErrMsg.EMPTY_INPUT);
+        }
+    }
+
+    public static void validateLength(List<String> lottoNumbers) {
+        if(lottoNumbers.size()!= Rule.LOTTO_NUMBER_LENGTH){
+            throw new IllegalArgumentException(ErrMsg.WRONG_LENGTH);
+        }
+    }
+
+    public static void validateDuplication(List<String> lottoNumbers) {
+        Set lottoNumberSet = new HashSet(lottoNumbers);
+        if(lottoNumberSet.size()!=lottoNumbers.size()){
+            throw new IllegalArgumentException(ErrMsg.DUPLICATED_INPUT);
+        }
+    }
+
+    public static void validateIntRange(int number) {
+        if(number< Rule.MIN_NUM || number>Rule.MAX_NUM){
+            throw new IllegalArgumentException(ErrMsg.WRONG_RANGE);
+        }
+    }
+
+    public static void validateTicketLength(List<Integer> ticket) {
+        if(ticket.size()!= Rule.LOTTO_NUMBER_LENGTH){
+            throw new IllegalArgumentException(ErrMsg.WRONG_LENGTH);
+        }
+    }
```
유효성 검사 로직들을 InputValidation에 모아둔 이유가 있을까요?

geunhwanlee commented on src/main/java/step3/model/value/ErrMsg.java:
```diff
@@ -0,0 +1,11 @@
+package step3.model.value;
+
+public class ErrMsg {
```
메시지를 상수로 만들어주신점 좋습니다!👍🏻

geunhwanlee replied:
인스턴스 생성을 방지하기 위해 private 생성자를 만드는건 어떨까요?

geunhwanlee commented on src/main/java/step3/model/value/OperationMsg.java:
```diff
@@ -0,0 +1,9 @@
+package step3.model.value;
+
+public class OperationMsg {
```
인스턴스 생성을 방지하기 위해 private 생성자를 만드는건 어떨까요?

geunhwanlee commented on src/main/java/step3/model/value/Rule.java:
```diff
@@ -0,0 +1,11 @@
+package step3.model.value;
+
+public class Rule {
```
인스턴스 생성을 방지하기 위해 private 생성자를 만드는건 어떨까요?

geunhwanlee commented on src/main/java/step3/model/value/Rule.java:
```diff
@@ -0,0 +1,11 @@
+
+public class Rule {
+
+    public static final int MAX_NUM = 45;
+    public static final int MIN_NUM = 1;
```
MAX_NUM, MIN_NUM 만 가지고는 어떤 의미인지 파악하기 어렵네요😭
더 구체적인 이름을 지어보는건 어떨까요?

geunhwanlee commented on src/main/java/step3/view/InputView.java:
```diff
@@ -0,0 +1,21 @@
+public class InputView {
+
+    public static String requestInputMoney() {
+        Scanner scanner = new Scanner(System.in);
```
Scanner를 정적 필드로 만들어둬도 좋을것 같네요.

badaelephant replied:
와 이부분도 생각을 하긴 했는데 적용을 못했었네요, 알려주셔서 감사해요

geunhwanlee commented on src/main/java/step3/view/InputView.java:
```diff
@@ -0,0 +1,21 @@
+    public static String requestInputLotto() {
+        Scanner scanner = new Scanner(System.in);
+        System.out.println(OperationMsg.REQUEST_INPUT_LOTTO);
+        return scanner.next();
```
`1, 2, 3, 4, 5, 6`을 입력하게 되면
`scanner.next()`는 어떤 값을 읽게될까요?

badaelephant replied:
next의 경우 공백으로 구분이 되어서 값이 잘못 읽히더라구요. 알려주셔서 감사합니다. ㅎㅎ

geunhwanlee commented on src/main/java/step3/view/InputView.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Scanner;
+import step3.model.value.OperationMsg;
+
+public class InputView {
```
인스턴스 생성을 방지하기 위해 private 생성자를 만드는건 어떨까요?

badaelephant replied:
프라이벳 생성자로 알려주신 내용은 잘 읽어보았고, 수정하였습니다. 감사합니다.

geunhwanlee commented on src/main/java/step3/view/OutputView.java:
```diff
@@ -0,0 +1,28 @@
+import step3.model.lotto.Tickets;
+import step3.model.machine.Result;
+
+public class OutputView {
```
인스턴스 생성을 방지하기 위해 private 생성자를 만드는건 어떨까요?

geunhwanlee commented on src/main/java/step3/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,55 @@
+
+    private List<LottoNumber> createLotto(List<String> lottoNumbers) {
+        InputValidation.validateLength(lottoNumbers);
+        InputValidation.validateDuplication(lottoNumbers);
```
`splitLottoNumbers`에서 중복을 제거한 값을 넘겨주기 때문에
중복체크 검사가 이루어지지 않을 것 같네요.🤔

badaelephant replied:
수정했습니다.

geunhwanlee reviewed (changes requested):
3단계 미션 잘 해주셨습니다!👍🏻

몇가지 코멘트를 드렸어요.
한번 생각해 보시고 재 리뷰요청 보내주세요!🙏

Commit:
refactor: 피드백 반영 - 유틸, 뷰, 상수관련
변경사항 1. 굳이 인스턴스를 생성할 필요가 없는 객체의 경우 private 생성자를 적용한다.
변경사항 2. 상수 파라미터의 경우 그 목적을 명확하게 알수있도록 한다.
변경사항 3. 공통 밸리데이션 로직은 가능하면 공통 목적의 클래스는 묶고(Ticket, Lotto) 도메인 클래스 내부에서 해결하도록 한다.
변경사항 4. View 클래스는 도메인 계층에 넣지 않는다.

Commit:
refactor: 통계 클래스를 외부로 빼서 기능을 분리한다.

Commit:
refactor: 주문 티켓 갯수 원시값을 Order로 포장한다.

Commit:
refactor: 티켓, 로또 등에서 결과를 생성해내는 대신, 결과관련 기능은 결과 관련 클래스로 완전 이관한다.

Commit:
refactor: print 관련하여, 상세 형태 잡는 부분을 OutputView로 위임한다.
result에서는 콤마로 구분된 값을 전달한다. (더 좋은 방법 고민)

Commit:
refactor: 유사한 역할을 하는 Ticket 클래스를 Lotto로 합침
기존의 분리 원인 - Integer 리스트로 Lotto를 만드는 것과, String 으로 Lotto를 만드는 케이스가 달라 클래스 자체를 분리하였음.
로또의 자동생성과 매뉴얼(input기반) 생성 기능을 담당하는 generator를 분리하여 Lotto와 Ticket을 함침.

Commit:
refactor: 테스트 코드 수정
로또, 티켓 합침
로또머신 테스트코드 생성
결과값 기능 독립

Commit:
style: 사용하지 않는 import 문 삭제

geunhwanlee commented on src/main/java/step3/controller/LottoController.java:
```diff
@@ -0,0 +1,77 @@
+    private Order makeOrder() {
+        Money money = getMoneyInput();
+        int ticketCount = money.availableTicketCount(Rule.TICKET_PRICE);
+        Order order =  new Order(ticketCount);
```
Order라는 객체가 꼭 필요할까요?🤔

geunhwanlee commented on src/main/java/step3/model/lotto/LottoList.java:
```diff
@@ -0,0 +1,23 @@
+        this.lottoList = lottoList;
+    }
+
+    public List<Integer> getResults(Lotto winningLotto) {
```
List\<Result\> 또는 Results를 리턴하면 더 좋겠네요!

geunhwanlee commented on src/main/java/step3/model/machine/LottoMachine.java:
```diff
@@ -0,0 +1,30 @@
+    public List<Lotto> issueAutoLottoList(Order order) {
+        List<Lotto> lottoList = new ArrayList<>();
+        int ticketCount=0;
+        while(order.leftToTicketing(ticketCount)){
```
수량만큼 for 문을 사용하는게 더 이해하기 쉽고 깔끔한 코드가 될것 같아요!

badaelephant replied:
이부분은 제가 getter를 최대한 안쓰려고 하다보니 그런것인데요, 저도 하면서 조금 억지스럽다고 생각을 했습니다. 혹시 getter, setter를 직접 사용하지 않는 것이 좋다고 얘기는 들었는데 그래도 허용되는 기준 같은 것이 있을까요? 예를들면 단순한 원시값의 경우 괜찮다던지.

geunhwanlee commented on src/main/java/step3/model/machine/LottoMachine.java:
```diff
@@ -0,0 +1,30 @@
+        }
+        return lottoList;
+    }
+    public List<Integer> createWinningLotto(String input) {
```
Lotto를 리턴하는게 더 좋을것 같아요!

geunhwanlee commented on src/main/java/step3/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,52 @@
+    public int getMatchCount(Lotto winningLotto){
+        return (int) lotto.stream().filter(winningLotto::isMatched).count();
+    }
+    public boolean isAllMatch(Lotto inputLotto){
```
사용하지 않는 메서드는 제거해주세요.😀

geunhwanlee commented on src/main/java/step3/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,52 @@
+        return this.lotto.stream().allMatch(inputLotto::isMatched);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof Lotto)) {
+            return false;
+        }
+        Lotto that = (Lotto) o;
+        Collections.sort(this.lotto);
+        Collections.sort(that.lotto);
+        return this.lotto.equals(that.lotto);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(lotto);
+    }
```
Lotto에서 equals, hashCode를 오버라이딩 할 필요는 없을것 같아요!
Lotto의 번호들이 모두 같으면 같은 Lotto일까요?🤔

geunhwanlee commented on src/main/java/step3/model/machine/LottoManualGenerator.java:
```diff
@@ -0,0 +1,38 @@
+        return lottoNumbers;
+    }
+
+    private void verifyLotto(List<Integer> lottoNumbers) {
```
로또 숫자를 검증하는건 Lotto에 있으면 더 좋겠네요!

badaelephant replied:
아 맞습니다. 수정하는게 좋겠네요

geunhwanlee reviewed (approved):
안녕하세요 용석님!

3단계 미션 잘해주셨어요👍🏻

이번 단계는 여기서 머지하도록 할게요.

몇가지 코멘트를 드렸으니 다음 단계에서 검토해주세요!😀
